aten/src/ATen/native/ForeachUtils.h

```diff
@@ -164,17 +168,26 @@ bool can_use_fast_route(TensorList tensors1, TensorList tensors2) {
 #endif
 }
 
-bool can_use_fast_route(TensorList tensors1, TensorList tensors2, TensorList tensors3) {
+bool can_use_fast_route(TensorList tensors1, TensorList tensors2, TensorList tensors3, Scalar scalar) {
+  TORCH_CHECK(tensors1.size() > 0, "Tensor list must have at least one tensor.");
+  TORCH_CHECK(tensors1.size() == tensors2.size(), "Tensor lists must have the same number of tensors, got ", tensors1.size(), " and ", tensors2.size());
+  TORCH_CHECK(tensors2.size() == tensors3.size(), "Tensor lists must have the same number of tensors, got ", tensors2.size(), " and ", tensors3.size());
+
 #ifdef __HIP_PLATFORM_HCC__
   return false;
 #else
   auto expected_device = tensors1[0].device();
+  auto expected_dtype = tensors1[0].dtype();
+
   for (int64_t i = 0; i < tensors1.size(); i++) {
     TORCH_CHECK(tensors1[i].sizes() == tensors2[i].sizes(), 
                 "Corresponding tensors from tensor lists have different sizes, got ", tensors1[i].sizes(), " and ", tensors2[i].sizes());
 
     TORCH_CHECK(tensors1[i].sizes() == tensors3[i].sizes(), 
                 "Corresponding tensors from tensor lists have different sizes, got ", tensors1[i].sizes(), " and ", tensors3[i].sizes());
+    TORCH_CHECK(tensors1[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors2[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
+    TORCH_CHECK(tensors3[i].dtype() == expected_dtype, "All tensors in the tensor list must have the same dtype.");
```

Comment:
It feels a little weird to me that we're doing some error checking in a function named "can_use_fast_route": I'd expect "can_use_fast_route" to just return a True/False and never error out.

Reply:
That's because can_use_fast_route is now doing double duty of `check_foreach_api_restrictions` and actually checking fast route. I agree it's a bit confusing, @izdeby what's the motivation behind this?

Reply:
I know. but i dont wanna iterate over the lists twice where first time will be checking for the hard requirements and error out and 2nd time over slow/fast path.

Reply:
Im trying to reduce CPU time. This is the main goal.

Reply:
How different are the numbers? I wouldn't expect checking the dtypes for these things to add very much to the overall runtime

Reply:
@zou3519, @ngimel returned `check_foreach_api_restrictions` and just did some refactoring and cleanup.